### PR TITLE
[Local Testing] SNOW-1184278 Support file imports in stored procedures

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   This PR is essentially adding https://github.com/snowflakedb/snowpark-python/pull/1329 for `MockStoredProcedure`. The code is placed in different places because a UDF expression is executed in `calculate_expression` but a stored procedure is executed on its own by invoking the `__call__` method of `MockStoredProcedure`. This could be refactored better.

### CODE REVIEW

File: src/snowflake/snowpark/mock/_stored_procedure.py [hunk 1]
```diff
@@ -97,7 +116,56 @@ def __call__(
...
+        last_import_directory = sys._xoptions.get("snowflake_import_directory")
+        sys._xoptions["snowflake_import_directory"] = temporary_import_path.name
+
+        # Save a copy of module cache
+        frozen_sys_module_keys = set(sys.modules.keys())
+
+        def cleanup_imports():
+            for module_path in self.imports:
+                if module_path in sys.path:
+                    sys.path.remove(module_path)
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/_stored_procedure.py:131]: I see in a previous [PR](https://github.com/snowflakedb/snowpark-python/pull/1335/files) we track udf imports inside a session
`analyzer.session.udf._udf_import_directories[udf_name]`.

so it doesn't apply here?

File: src/snowflake/snowpark/mock/_stored_procedure.py [hunk 2]
```diff
@@ -116,37 +184,76 @@ def __call__(
...
+            elif file_extension == ".py":
+                module_root_dir = os.path.join(local_path, "..")
+            elif file_extension == ".zip":
+                module_root_dir = local_path
+            else:  # directory
+                module_root_dir = os.path.join(local_path, "..")
+
+            absolute_module_path = os.path.abspath(module_root_dir)
+            module_name = file_name.split(".")[
+                0
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/_stored_procedure.py:248]: looks like this part of code is exactly the same as the one in UDF: https://github.com/snowflakedb/snowpark-python/pull/1335/files

shall we make it a util function?

File: src/snowflake/snowpark/mock/_stored_procedure.py [hunk 3]
```diff
@@ -189,34 +296,49 @@ def _do_register_sp(
             anonymous,
         )
 
-        if is_pandas_udf:
-            raise TypeError("pandas stored procedure is not supported")
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/_stored_procedure.py:193]: do we need to keep this? in the live code it still raises the error

File: src/snowflake/snowpark/mock/_stored_procedure.py [hunk 4]
```diff
@@ -189,34 +296,49 @@ def _do_register_sp(
...
+                    self._import_file(local_path, import_path, sproc_name=sproc_name)
+
+        if type(func) is tuple:  # register from file
+            if sproc_name not in self._sproc_level_imports:
+                self._sproc_level_imports[sproc_name] = set()
+            module_name = self._import_file(func[0], sproc_name=sproc_name)
+            func = (module_name, func[1])
+
+        if sproc_name in self._sproc_level_imports:
+            sproc_imports = self._sproc_level_imports[sproc_name]
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/_stored_procedure.py:328]: I'm a bit curious about the `if sproc_name in self._sproc_level_imports:` logic here.
is this for re-registering a stored proc with the same name?

File: src/snowflake/snowpark/mock/_stored_procedure.py [hunk 5]
```diff
@@ -97,7 +116,56 @@ def __call__(
...
+            # Restore snowflake_import_directory
+            if last_import_directory is not None:
+                sys._xoptions["snowflake_import_directory"] = last_import_directory
+            else:
+                del sys._xoptions["snowflake_import_directory"]
+
+        with ExitStack() as stack:
+            stack.callback(cleanup_imports)
+
+            for module_path in self.imports:
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/_stored_procedure.py:150]: similar to my question above, in UDF implementation we have the `if udf_name in analyzer.session.udf._udf_level_imports:` check, but seems like we remove that concept in sp.

want to know why the diff

File: src/snowflake/snowpark/mock/_stored_procedure.py (hunk 1 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/_stored_procedure.py:131]: I've removed the logic around import directories in both UDF and stored proc implementation, there's no need to track them.

File: src/snowflake/snowpark/mock/_stored_procedure.py (hunk 5 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/_stored_procedure.py:150]: `if udf_name in analyzer.session.udf._udf_level_imports:` is the same as `if sproc_name in self._sproc_level_imports:` below. UDF/Sproc level imports overrides session level imports. If no UDF/Sproc level imports specified, session level imports are used.

File: src/snowflake/snowpark/mock/_stored_procedure.py (hunk 4 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/_stored_procedure.py:331]: Ditto.

**Review by sfc-gh-jrose — APPROVED** [review]

**Review by sfc-gh-aling — APPROVED** [review]
LGTM, let's fix the test failure and merge the PR

**Pull request merged**